crates/sui-graphql-rpc/src/types/suins_registration.rs

```diff
@@ -350,7 +429,95 @@ impl NameService {
             .to_rust()
             .ok_or_else(|| Error::Internal("Malformed Suins Domain".to_string()))?;
 
-        Ok(Some(field.value))
+        let domain = Domain(field.value);
+
+        // We attempt to resolve the domain to a record, and if it fails, we return None. That way
+        // we can validate that the name has not expired and is still valid.
+        let Some(_) = Self::resolve_to_record(ctx, &domain, checkpoint_viewed_at).await? else {
+            return Ok(None);
+        };
+
+        Ok(Some(domain.0))
+    }
+
+    /// Query for a domain's NameRecord, its parent's NameRecord if supplied, and the timestamp of
+    /// the checkpoint bound.
+    async fn query_domain_expiration(
+        db: &Db,
+        page: Page<object::Cursor>,
+        record_id: SuiAddress,
+        parent_record_id: Option<SuiAddress>,
+        checkpoint_viewed_at: Option<u64>,
+    ) -> Result<Option<DomainExpiration>, Error> {
+        let mut object_ids = vec![record_id];
+        let mut name_record = None;
+        let mut parent_name_record = None;
+
+        if let Some(id) = parent_record_id {
+            object_ids.push(id);
+        }
+
+        // prepare the filter for the query.
+        let filter = ObjectFilter {
+            object_ids: Some(object_ids),
+            ..Default::default()
+        };
+
+        let response = db
+            .execute_repeatable(move |conn| {
+                let Some((lhs, rhs)) = consistent_range(conn, checkpoint_viewed_at)? else {
+                    return Ok::<_, diesel::result::Error>(None);
+                };
+
+                let timestamp_ms = Checkpoint::query_timestamp(conn, rhs as i64)?;
+
+                let sql = build_objects_query(
+                    crate::consistency::View::Consistent,
+                    lhs as i64,
+                    rhs as i64,
+                    &page,
+                    move |query| filter.apply(query),
+                    move |newer| newer,
+                );
+
+                let objects: Vec<StoredHistoryObject> =
+                    conn.results(move || sql.clone().into_boxed())?;
+
+                Ok(Some((timestamp_ms, objects)))
+            })
+            .await?;
+
+        let Some((timestamp_ms, results)) = response else {
+            return Err(Error::Client(
+                "Requested data is outside the available range".to_string(),
+            ));
+        };
+
+        // Max size of results is 2. We loop through them, convert to objects, and then parse
+        // name_record. We then assign it to the variable based on the address.
```

Comment:
Because you rely on this here, I would construct the `Page` within the body of this function, rather than accept it as a parameter.

Reply:
yeah good call